src/logic/map_objects/tribes/tribe_descr.cc

```diff
@@ -546,4 +637,242 @@ DescriptionIndex TribeDescr::add_special_ware(const std::string& warename) {
 		throw GameDataError("Failed adding special ware '%s': %s", warename.c_str(), e.what());
 	}
 }
+
+void TribeDescr::process_productionsites(const World& world) {
+	// Get a list of productionsites - we will need to iterate them more than once
+	std::set<ProductionSiteDescr*> productionsites;
```

Comment:
Never use `std::set<SomePointer*>` in logic code. Danger of desyncs!

Reply:
Desync confirmed. Will switch to DescriptionIndex instead.

Reply:
`std::set<DescriptionIndex>` is desync-proof, so there must be another desync coming from somewhere else then…

Reply:
I was testing against a master replay, maybe there's already enough difference there to make it incompatible.

The strings make for less lines of code, so let's go with them anyway for now. We can revisit the decision once I have the AI part programmed.

Reply:
Tested and working.
There are still some std::sets of pointers left in this function; once those are gone this is good to go in.

Reply:
The pointers are there for temporary calculation during configuration only. I don't see how they can affect the game state at all, since everything that might possibly be game-state related (ProductionSite Descr members) are now descr().name().

Replacing the pointers will result in lots of superfluous "fetch-the-same-object-again-and-dynamic-cast".

Reply:
Just about everything that is done in the logic may, in sub-sub-sub-sub-function calls, write a syncstream entry. If the same entries are written in a different order, the game will believe that there has been a desync. But according to the reference the definition of `std::set` is
```
template<
    class Key,
    class Compare = std::less<Key>,
    class Allocator = std::allocator<Key>
> class set;
```
so perhaps something like this would work?
```
std::set<ProductionSite*, struct {
	bool operator()(const ProductionSite*& a, const ProductionSite*& b) const {
    	return a->descr().name().compare(b->descr().name()) > 0;
	}
}>
```

Reply:
No, it would not work. You are defining a sort order for a `ProductionSite`, but we are dealing with `ProductionSiteDescr` here. I would agree if we were calculating this within a `ProductionSite`, but we're not. All these calculations happen before the game state is loaded, and the results are written as strings into the `ProductionSiteDescr`, so this can't possibly affect the game state.

Reply:
Argh… updated the comment to make it work for `ProductionSiteDescr`

Reply:
This doesn't change the fact that this can't possibly affect the game state though:

* Pointers are only used temporarily during tribes configuration and get destroyed before the game starts loading
* Anything visible by the game state (`ProductionSiteDescr` members) is written as strings, not as pointers, so your proposed sorting algorithm has 0 impact

Reply:
All right. Please add a comment explaining this though.
```suggestion
	// WARNING: Using pointers as keys in sets and maps is dangerous!
	// It is fine in this function though because all we do while iterating them
	// is to insert strings into sets. If this function is ever changed in such
	// a way that the order of iteration matters, this will desync though.
	std::set<ProductionSiteDescr*> productionsites;
```
Code LGTM, tested and working.

Reply:
I have added a 1-liner comment.

I have also run a 4 hour multiplayer game, clang build vs. gcc build, and then watched the replays with both builds. No desyncs.